File: Algorithms/1. Two Sum/Two Sum.cs
//Author: Tushar Jaiswal
//Creation Date: 09/30/2015

/*Given an array of integers, find two numbers such that they add up to a specific target number.
The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.
You may assume that each input would have exactly one solution.
Input: numbers={2, 7, 11, 15}, target=9
Output: index1=1, index2=2
Input: numbers={-3,4,3,90}, target=0
Output: index1=1, index2=3*/

public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        if(nums.Length == 2)
        {
            return new int[]{1, 2};
        }
        else
        {
            int index;
            Dictionary<int, int> inputIntegers = new Dictionary<int, int>();
            for (int i = 0; i <nums.Length; i++)//Add all valid integers to a dictionary except two same numbers
            {
                try
                {
                    inputIntegers.Add(nums[i], i);
                }
                catch (Exception ex)
                {
                    if(inputIntegers.TryGetValue(nums[i], out index))//If this number has already been added to dictionary, check to see if adding these two numbers gets us the target
                    {
                        if(index != i && nums[i]*2 == target)
                        {
                            return(new int[]{index+1, i+1});
                        }
                    }
                }
            }
            
            for (int i=0; i < nums.Length; i++)
            {
                //Check if for the current integer we can find the desired integer amongst the valid integers stored in the dictionary
                if(inputIntegers.TryGetValue((target-nums[i]), out index))
                {
                    if(index != i)//If twice of current number is the target but there isn't any duplicate number in the array, this check prevents false indexes from being returned
                    {
                        return new int[]{i+1, index+1};   
                    }
                }
            }
            
            return null;
        }
    }
}

File: Algorithms/1. Two Sum/TwoSum.py
# Author: Tushar Jaiswal
# Creation Date: 2025-08-05

# Problem Source: https://leetcode.com/problems/two-sum

# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

# You may assume that each input would have exactly one solution, and you may not use the same element twice.

# You can return the answer in any order.

# Example 1:
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

# Example 2:
# Input: nums = [3,2,4], target = 6
# Output: [1,2]

# Example 3:
# Input: nums = [3,3], target = 6
# Output: [0,1]
 
# Constraints:
# 2 <= nums.length <= 10^4
# -10^9 <= nums[i] <= 10^9
# -10^9 <= target <= 10^9
# Only one valid answer exists.
 
# Follow-up: Can you come up with an algorithm that is less than O(n2) time complexity?

# Runtime complexity: O(n)
# Space complexity: O(n)

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        val_index_map = {}
        for i, num in enumerate(nums):
            if target - num in val_index_map:
                return [val_index_map[target - num], i]
            val_index_map[num] = i
        return []
